clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java

```diff
@@ -182,6 +216,180 @@ public CompletableFuture<Map<TopicPartition, OffsetAndTimestampInternal>> fetchO
                         result.fetchedOffsets));
     }
 
+    private boolean maybeCompleteWithPreviousKnownException(CompletableFuture<Boolean> result) {
+        Throwable exception = cachedUpdatePositionsException.getAndSet(null);
+        if (exception != null) {
+            // Return exception that may have been encountered on a previous attempt to update
+            // positions, after the triggering event had already expired.
+            result.completeExceptionally(exception);
+            return true;
+        }
+        return false;
+    }
+
+    public CompletableFuture<Boolean> updateFetchPositions(long deadlineMs) {
```

Comment:
This `CompletableFuture<Boolean>` is a slightly tricky thing. The boolean I think refers to whether we have all of the offset information we need. But there are a lot of "results" flying around and I'm finding it tricky to follow this boolean through the various methods. I wonder whether a specific variable name for the future which contains the flag which indicates whether we have all of the offset information used at all points whether it's currently a variable or a method argument would help my tiny brain.

Reply:
Yeap, agree is not straightforward. The boolean only indicates if the event `hasAllFetchPositions` when it starts (meaning that it didn't have to fetch offsets). The results flying around really have no direct impact that result (only indirect, setting a position, so next time the hasAllFetchPositions will find it). 

Definitely having specific var names for the result will help, going for it now...